cocotb/_sim_versions.py

```diff
@@ -6,39 +6,108 @@
 Classes to compare simulation versions.
 
 These are for cocotb-internal use only.
-
-.. warning::
-    These classes silently allow comparing versions of different simulators.
 """
 
 from distutils.version import LooseVersion
 
 
-class ActivehdlVersion(LooseVersion):
+class _LooseVersionTypeChecked(LooseVersion):
+    """Do a strict type check so that versions of different simulators cannot be compared.
+
+    Example:
+        >>> IcarusVersion("10.3 (stable)") == VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") != VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") > VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") >= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") < VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") <= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+    """
+
+    def __eq__(self, other):
+        if type(other) is not type(self):
+            raise TypeError(
+                "Comparing versions of different simulators doesn't make sense"
+            )
+        else:
+            return super().__eq__(other)
+
+    def __lt__(self, other):
+        if type(other) is not type(self):
+            raise TypeError(
```

Comment:
`return NotImplemented` might be just as good here

Reply:
I had tried `NotImplemented` before because that would also allow me to use `@total_ordering`, but it didn't actually raise an exception in the end but somehow resolved the comparison instead.

Reply:
Are you sure? Note that `NotImplemented` does _not_ work for `__eq__`, only the other comparison methods.

Reply:
Ah, ok, I did also use it for `__eq__`. I'll try later.

Reply:
Ok, please take another look - though I have to say it look more consistent/uniform before.